activerecord/lib/active_record/insert_all.rb

```diff
@@ -10,15 +10,20 @@ class InsertAll # :nodoc:
     def initialize(model, inserts, on_duplicate:, update_only: nil, returning: nil, unique_by: nil, record_timestamps: nil)
       raise ArgumentError, "Empty list of attributes passed" if inserts.blank?
 
-      @model, @connection, @inserts, @keys = model, model.connection, inserts, inserts.first.keys.map(&:to_s)
+      @model, @connection, @inserts = model, model.connection, inserts
       @on_duplicate, @update_only, @returning, @unique_by = on_duplicate, update_only, returning, unique_by
       @record_timestamps = record_timestamps.nil? ? model.record_timestamps : record_timestamps
 
       disallow_raw_sql!(on_duplicate)
       disallow_raw_sql!(returning)
 
+      if has_attribute_aliases?(@inserts.first)
+        resolve_attribute_aliases
+      end
```

Comment:
I think I would have put this check inside `resolve_attribute_aliases`. It's better for encapsulation.